### PR TITLE
Fix Gc.minor_{words,free} by removing [@@noalloc]

### PR DESCRIPTION
**Pull request by stedolan**
In native code, `Gc.minor_words` and `Gc.minor_free` return wrong data, with code like:

```ocaml
let a = Gc.minor_words () in
f ();
let c = Gc.minor_words () in
c -. a
````

returning 0 even if `f` allocates thousands of words. The issue is the `[@noalloc]` annotation on those functions: indeed, they do not allocate, but the `[@noalloc]` annotation prevents them seeing an accurate allocation pointer. By removing this annotation, they still don't allocate, but they return accurate data.

(I'm aware that the documentation says "This number is accurate in byte-code programs, but only an approximation in programs compiled to native code.", so one could argue that arbitrarily wrong answers are not a bug. Still, the results are badly wrong and easily fixed)

### CODE REVIEW

**Comment by chambart** [comment]
I effectively overlooked that when introducing it. good catch.

File: testsuite/tests/misc/gcwords.ml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  let c = Gc.minor_words () in
+  c -. a
+
+let () =
+  let n = measure (fun () -> a.(0) <- Gc.minor_words ()) in
+  (* Gc.minor_words should not allocate, although bytecode
+     generally boxes the floats *)
+  assert (n < 10.);
+  if Sys.backend_type = Sys.Native then assert (n = 0.);
+  let n = measure (fun () -> r := Some (allocate_lots 42 10)) in
```
Comment (chambart) [inline testsuite/tests/misc/gcwords.ml]: The right way to do that now is `Sys.opaque_identity`

Comment (stedolan) [reply testsuite/tests/misc/gcwords.ml]: Changed.

**Comment by xavierleroy** [comment]
Great catch-22, well spotted!  

Could you please put the Changes entry in the "4.05.0" section?  This is a bug fix and it should go both on trunk and on 4.05 branch.

**Comment by gasche** [comment]
Note that this may break library that re-export the `external` declaration, as the type-checker will complain if the `.ml` and `.mli` disagree of noalloc-ness of a primitive. (I think that for example Batteries may have to be patched.)

(I still agree that fixing this and having the fix in 4.05 is the good solution.)

For this reason we may or may not mark it as a "breaking change" in the Changelog. It also mean that doing this as early as possible in the 4.05 release process is important -- to give library authors time to update their code before the release.

**Comment by stedolan** [comment]
OK, Changes entry moved to 4.05 section. Should it be `-` or `*`?

@gasche  re-exporting external declarations seems fragile. Would it not be better for Batteries to export `val minor_words : unit -> float` instead? Or even to just do `include Gc`?

**Comment by gasche** [comment]
Re. `-` and `*`, I think either are correct choices. In general there is an unspoken agreement that "having to change already-buggy code does not mandate a `*` mark". Here, the broken code is not buggy, but one could agree that this is a very specific failure scenario. I think the purpose of `*` marks is to let users quickly review what risks breaking their code when reading the Changelog, so the main question I think is whether this particular mark would be useful to anyone. For Batteries, I already know, so it's not very important; it may or may not affect other people (other `external` declarations are used in libraries, cf. the `%string_set` questions, but I don't know of any library that would export this one in particular). Long story short, do what you think is best.

Re. Batteries: The implementation does `include Gc`, but for the interface we re-export. There are several reasons, for example:

- Having the interface of both Batteries-added functions and standard functions makes for an easier-to-browse documentation
- Generally we could argue that if exporting `val` was always the better design, then the standard library should do it as well. If there are concerns with doing so (for example, performance, although inlining should  solve that), the concerns probably also apply to Batteries. Mimicking stdlib is the safe bet here.
- Whenever possible we want to make sure that the `BatFoo` module always also respect the module type of the standard `Foo` module; this lets clients use it interchangeably, and it is also the way we do internal testing (we check that this test pass to detect new stdlib functions we should bind and possibly backport).

**Comment by damiendoligez** [comment]
The stdlib exports externals for historical reasons (concerns with the performance of byte-code programs). It is now considered a bad idea.

Still, this change is going to break the build for some libraries, so you should use a `*`.

**Comment by stedolan** [comment]
> Still, this change is going to break the build for some libraries, so you should use a *.

Done.

**Comment by gasche** [comment]
Merged in trunk ( 06396f0acf5bfb70e157aa75b6706d55dc018efc ) and 4.05 ( f5412927ba7b16a0f88b5af0f1f0edfb0ec091d6 ).

**Pull request merged**